File: 0056. Merge Intervals.java
// Medium
// Greedy
// Time:O(nlogn), Space:O(n)
// https://leetcode.cn/problems/merge-intervals/

import java.util.*;

class Solution {
    public int[][] merge(int[][] intervals) {
        List<int[]> res = new ArrayList<>();

        // Sort by start time
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        // Get first interval
        int start = intervals[0][0];
        int end = intervals[0][1];

        for (int i = 1; i < intervals.length; i++) {
            // If the cur start > prev end -> no merge
            if (intervals[i][0] > end) {
                res.add(new int[] { start, end }); // Put prev interval to res
                // update the interval to compare later
                start = intervals[i][0];
                end = intervals[i][1];
            } else {
                // merge -> update the end
                end = Math.max(end, intervals[i][1]);
            }
        }
        // add last interval
        res.add(new int[] { start, end });
        return res.toArray(new int[res.size()][]);
    }
}

File: 0001-0500/0056. Merge Intervals.py
# Time:O(nlogn), Space:O(n)

from typing import List

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if not intervals:
            return []
        
        # sort by start
        intervals.sort(key=lambda x: x[0])

        res = []
        cur_start, cur_end = intervals[0]

        for start, end in intervals[1:]:
            # overlap
            if start <= cur_end:
                cur_end = max(cur_end, end)
            else:
                res.append([cur_start, cur_end])
                cur_start, cur_end = start, end
        
        res.append([cur_start, cur_end]) # last interval
        return res
